File: LeetCode/876-middle-of-the-linked-list/Solution.java
class Solution {
  /**
   * Returns the middle node of a singly linked list using the two-pointer technique.
   * This method employs a slow and fast pointer: - The slow pointer advances one node per iteration. - The fast pointer advances two nodes per iteration. When the fast pointer reaches the end of the list, the slow pointer will be at the middle.
   * If the list has an even number of nodes, the method returns the second middle node.
   *
   * @param head the head of the singly linked list
   * @return the middle node of the list (includes all subsequent nodes)
   */
  public ListNode middleNode(ListNode head) {
    ListNode slow = head;
    ListNode fast = head;

    while (fast != null && fast.next != null) {
      slow = slow.next;
      fast = fast.next.next;
    }

    return slow;
  }
}

File: LeetCode/876-middle-of-the-linked-list/solution.ts
/**
 * Represents a node in a singly linked list.
 */
class ListNode {
  val: number;
  next: ListNode | null;

  /**
   * Constructs a new ListNode.
   * @param val - The integer value stored in the node. Defaults to 0.
   * @param next - Reference to the next node in the list. Defaults to null.
   */
  constructor(val?: number, next?: ListNode | null) {
    this.val = val ?? 0;
    this.next = next ?? null;
  }
}

/**
 * Returns the middle node of a singly linked list using the two-pointer approach.
 *
 * This function uses two traversal pointers:
 * - `slow` advances one node per iteration.
 * - `fast` advances two nodes per iteration.
 *
 * When `fast` reaches the end of the list, `slow` will be at the middle.
 * If the list has an even number of nodes, the second middle node is returned.
 *
 * @param head - The head of the singly linked list.
 * @returns The middle node of the list, or null if the input is null.
 */
function middleNode(head: ListNode | null): ListNode | null {
  let slow: ListNode | null = head;
  let fast: ListNode | null = head;

  while (fast != null && fast.next != null) {
    slow = slow!.next;
    fast = fast.next.next;
  }

  return slow;
};